File: dcp-solutions/2024-10-22/solution.ex
```elixir
defmodule MaxStack do
  defstruct stack: [], max_stack: []

  @spec push(Any) :: :ok
  def push(val), do: {:ok, %{stack: [val | stack], max_stack: max_stack([val])}}

  @spec pop() :: Any | nil
  def pop() do
    case Enum.at(stack, -1) do
      nil -> :error
      nil -> nil
      val when val in max_stack do
        {:ok, %{stack: [], max_stack: []}}
      else
        {:ok, %{stack: stack, max_stack: [val | max_stack]}}
      end
    end
  end

  @spec max() :: Any | nil
  def max(), do: max_stack()
end

defmodule MaxStack.MaxStack do
  @spec find_max(list) :: non_neg_integer | nil
  def find_max([]), do: nil

  @spec find_max([non_neg_integer]) :: non_neg_integer | nil
  def find_max([^max, tail]), do: max

  defp find_max([val | tail]), do: val > max ? val : max end

  defp find_max([^_, tail]), do: find_max(tail)
end
```

File: dcp-solutions/2024-10-22/solution.go
```go
package stack

import (
	"errors"
)

type MaxStack struct {
	stack []int
	max   []int
}

func NewMaxStack() *MaxStack {
	return &MaxStack{
		stack:  make([]int, 0),
		max:    make([]int, 0),
	}
}

func (s *MaxStack) Push(val int) {
	s.stack = append(s.stack, val)
	if len(s.max) == 0 || s.max[len(s.max)-1] <= val {
		s.max = append(s.max, val)
	}
}

func (s *MaxStack) Pop() (int, error) {
	if len(s.stack) == 0 {
		return 0, errors.New("stack is empty")
	}

	val := s.stack[len(s.stack)-1]
	s.stack = s.stack[:len(s.stack)-1]

	if val == s.max[len(s.max)-1] {
		s.max = s.max[:len(s.max)-1]
	}
	return val, nil
}

func (s *MaxStack) Max() (int, error) {
	if len(s.max) == 0 {
		return 0, errors.New("stack is empty")
	}

	return s.max[len(s.max)-1], nil
}
```